### ISSUE TITLE
broken should_receive expectations with do blocks don't cause the test to go red on classes which handle exceptions

### ISSUE THREAD

**Issue by froodian** [rspec/rspec-mocks#203]
If you have some code which rescues from exceptions, and you use the `should_receive(:foo) do |arg|` syntax to expect that class to make a call to a mock, if the assertions inside the do block fail, the test doesn't go red as expected.

Test-case which demonstrates the bug:

``` ruby
require 'spec_helper'

module BugReport203
  describe "an expectation on a class with a rescue block" do
    class TestClass
      def test_dependency
        @dependency ||= TestDependency.new()
      end

      def test_method
        begin
          self.test_dependency.some_call("I'm calling this method with unexpected arguments - THIS SHOULD CAUSE THE TEST TO GO RED!  But doesn't.")
        rescue
          # Exception-handling logic here
        end
      end
    end

    class TestDependency
      def some_call(arguments)

      end
    end

    # This test should be red!
    it "should know the method was called with wrong arguments" do
      mock_dependency = mock(TestDependency)

      test_object = TestClass.new()
      test_object.stub(:test_dependency).and_return(mock_dependency)

      # Obviously in this trivial case of checking string equality, I could use
      #
      # mock_dependency.should_receive(:some_call).with("expected arguments")
      #
      # which will go red as expected.
      # This bug applies to the more complex case where I actually need the do block to make my assertions
      mock_dependency.should_receive(:some_call) do |arg|
        arg.should == "expected arguments"
      end

      test_object.test_method()
    end
  end
end
```

**Comment by jooshbzm** [comment]
We ran into this problem as well. To work around it, we just handled exception in a method, then monkeypatched that method in spec/support. It's working well for now.

Something like this:

```
# my_class.rb
class MyClass
  def my_method
    begin
      do_something
    rescue Exception => e
      handle_exception(e)
    end
  end

  def handle_exception(e)
    do_something_else
  end
end
```

```
# spec/support/helpers.rb

class MyClass
  original_handle_exception = instance_method(:handle_exception)

  define_method(:handle_exception) do |e|
    raise e if e.is_a?(RSpec::Mocks::MockExpectationError)

    original_handle_exception.bind(self).call(e)
  end
end
```

**Comment by myronmarston** [comment]
Another solution: rspec 3.3 (to be released ASAP) contains a new `aggregate_failures` feature that allows multiple failures within a single example and raises a single exception at the end.

**Comment by jarmo** [comment]
Why didn't that `throw/catch` solution suit that @moll suggested above? Seems to be most natural way of fixing this problem and then you could say something in the lines of `all other testing frameworks are fragile` :)

**Comment by myronmarston** [comment]
@jarmo -- changing to `throw/catch` would be a breaking change to the API, and rspec-mocks and rspec-expectations are both usable with other test frameworks (e.g. cucumber and minitest) which expect the current behavior (raising an exception).  The best we can do for now is make it configurable.

...which, as of RSpec 3.3 it is :).  It's not a documented public API (and therefore may have breaking changes in later 3.x releases) but there's now this:

``` ruby
RSpec::Support.failure_notifier = lambda do |error, options|
  # do something with the error
end
```

So I believe you could do this:

``` ruby
RSpec::Support.failure_notifier = lambda do |error, options|
  throw :rspec_failure, error
end

RSpec.configure do |c|
  c.around do |ex|
    error = catch(:rspec_failure) do
      ex.run
      nil
    end

    raise error if error
  end
end

RSpec.describe "Rescuing exception failures" do
  it "does not prevent it from failing" do
    begin
      expect(1).to eq(2)
    rescue Exception
    end
  end

  it 'allows passing examples to pass' do
    expect(1).to eq(1)
  end
end
```

**Comment by jarmo** [comment]
@myronmarston: instead of "fixing" bad design (tests not failing due to bad `rescue Exception` clause smells) you see that adding new features/configuration parameters is a better solution?

I hope RSpec is not heading Java's way where backwards compatibility is the most important thing to have so all that crap from it's beginning still exists.

**Comment by myronmarston** [comment]
> I hope RSpec is not heading Java's way where backwards compatibility is the most important thing to have so all that crap from it's beginning still exists.

We made _lots_ of backwards incompatible changes when we released 3.0.  4.0 will include more backwards incompatible changes (although much fewer, I believe -- and no major syntax changes).

We do follow SemVer, which means we've promised to our users not to intentionally break backwards compatibility in a minor or patch release, so making that kind of change in a 3.x release simply isn't an option.

All that said: I plan to think more about this to see if there's a backwards-compatible way we can do this in a 3.x release.

The snippet I pasted above was simply meant to show that for people who want this sooner, there is a way to do it.

**Comment by jarmo** [comment]
Actually, I didn't think that this change as a backwards-compatible change. I'd be really fine if that would end up in 4.x series instead of not happening ever and being broken in future major versions as well.

I know what SemVer is and I'm following that with my own OSS libraries as well.

Thank You for all the good work so far!
